### PR TITLE
Add option to log to file

### LINKED ISSUE #795: Support logging to file
This is mostly for not-operators that still depend on op-rs for consistency with the rest of the platform (like the [OPA bundle builder](https://github.com/stackabletech/opa-operator/tree/a1618d5490d29841d804e9be332f1a4c829a6385/rust/bundle-builder)).

It should support rotation via https://docs.rs/tracing-appender/latest/tracing_appender/.

### PR DESCRIPTION
**Pull request by nightkr**
# Description

Fixes #795

## Definition of Done Checklist

- Not all of these items are applicable to all PRs, the author should update this template to only leave the boxes in that are relevant
- Please make sure all these things are done and tick the boxes

```[tasklist]
# Author
- [x] Integration tests passed (for non trivial changes)
```

```[tasklist]
# Reviewer
- [x] Code contains useful comments
- [x] Changelog updated
- [x] Cargo.toml only contains references to git tags (not specific commits or branches)
```

```[tasklist]
# Acceptance
- [ ] Feature Tracker has been updated
- [ ] Proper release label has been added
```

### CODE REVIEW

File: crates/stackable-operator/src/product_logging/framework.rs
Comment (nightkr) [inline crates/stackable-operator/src/product_logging/framework.rs:758]: https://github.com/stackabletech/opa-operator/pull/578 migrates to this mechanism, see https://github.com/stackabletech/opa-operator/pull/578#discussion_r1611798885

**Comment by nightkr** [comment]
One downside here is that `tracing-appender` rotates based on time rather than disk space, so we can no longer provide strict guarantees that the logs will never outgrow their respective volumes... There are PRs up for addressing this already (see https://github.com/tokio-rs/tracing/issues/2972: https://github.com/tokio-rs/tracing/pull/2904, https://github.com/tokio-rs/tracing/pull/2497) but none that have been merged yet.

**Comment by sbernauer** [comment]
Well that's a pity :/

File: crates/stackable-operator/src/logging/mod.rs
Comment (sbernauer) [inline crates/stackable-operator/src/logging/mod.rs:30]: > This file will be rotated regularly

Actually it looks like it defaults to no rotation: https://github.com/tokio-rs/tracing/blob/ba387ddc03f17eb91e70efe58ae007326ecb9cd4/tracing-appender/src/rolling/builder.rs#L41

Can we please explicitly set a rotation?

**Review by sbernauer — COMMENTED** [review]
Only looked at the Rust code

File: crates/stackable-operator/src/logging/mod.rs
Comment (nightkr) [reply crates/stackable-operator/src/logging/mod.rs:30]: Argh, you're right. It _used to be_ hourly, but I screwed up when migrating to the builder API instead.

Comment (nightkr) [reply crates/stackable-operator/src/logging/mod.rs:30]: 2c5e3548b0437bb9d6c9ff71c783c8a328fdd20c

Comment (sbernauer) [reply crates/stackable-operator/src/logging/mod.rs:30]: You "just" forget to push the commit ;)

Comment (nightkr) [reply crates/stackable-operator/src/logging/mod.rs:30]: GODDAMNIT IT HAPPENED AGAIN

Comment (nightkr) [reply crates/stackable-operator/src/logging/mod.rs:30]: 2c5e3548b0437bb9d6c9ff71c783c8a328fdd20c

**Comment by sbernauer** [comment]
@siegfriedweber would you be willing to review this PR? I think you are the best person to do so, as you can best asses the `so we can no longer provide strict guarantees that the logs will never outgrow their respective volumes` problem. Maybe this is a blocker (given customers already had problems with the volume sizes), maybe not.

Also you know the vector transformation far better

**Review by siegfriedweber — CHANGES_REQUESTED** [review]
Only minor things. The transform rule is implemented very well.

File: crates/stackable-operator/CHANGELOG.md
Comment (nightkr) [reply crates/stackable-operator/CHANGELOG.md:1]: fecbf885c419177b579b2c95272b92cdf4f99454

File: crates/stackable-operator/src/product_logging/framework.rs
Comment (nightkr) [reply crates/stackable-operator/src/product_logging/framework.rs]: 31d8fea9717ed3001a55ccb950168bc1aca4e488

File: crates/stackable-operator/CHANGELOG.md
Comment (siegfriedweber) [inline crates/stackable-operator/CHANGELOG.md:18]: This entry must also be moved to the unreleased section.

File: crates/stackable-operator/src/logging/mod.rs
Comment (siegfriedweber) [inline crates/stackable-operator/src/logging/mod.rs:44]: `app_name` should probably be the filename prefix:

```suggestion
            .filename_prefix(app_name.to_string())
            .filename_suffix("tracing-rs.json")
```

This generates filenames like `opa-bundle-builder.2024-07-25-07.tracing-rs.json`.

**Review by siegfriedweber — CHANGES_REQUESTED** [review]

File: crates/stackable-operator/CHANGELOG.md
Comment (nightkr) [reply crates/stackable-operator/CHANGELOG.md]: >.< 43d3baec1c670926ca24d9bb92a24eac3f074970

File: crates/stackable-operator/src/logging/mod.rs
Comment (nightkr) [reply crates/stackable-operator/src/logging/mod.rs:44]: I agree on a level, but this would be a behaviour change that I'm not sure how Vector reacts to. Previously, Vector would only have the "current" file listed for ingestion. With this change, it would also see all already-rotated files.

Comment (siegfriedweber) [reply crates/stackable-operator/src/logging/mod.rs:44]: My suggested change is actually only a cosmetic one. With the change, the filename is `opa-bundle-builder.2024-07-25-07.tracing-rs.json` and without it, it is `2024-07-25-07.opa-bundle-builder.tracing-rs.json`. This makes no difference in processing.

The current implementation creates a new file every hour. Vector starts watching it and still watches the old ones, which is no problem:

```
Found new file to watch. file=/stackable/log/bundle-builder/2024-07-25-11-52.opa-bundle-builder.tracing-rs.json
Found new file to watch. file=/stackable/log/bundle-builder/2024-07-25-11-53.opa-bundle-builder.tracing-rs.json
Found new file to watch. file=/stackable/log/bundle-builder/2024-07-25-11-54.opa-bundle-builder.tracing-rs.json
Found new file to watch. file=/stackable/log/bundle-builder/2024-07-25-11-55.opa-bundle-builder.tracing-rs.json
Found new file to watch. file=/stackable/log/bundle-builder/2024-07-25-11-56.opa-bundle-builder.tracing-rs.json
Found new file to watch. file=/stackable/log/bundle-builder/2024-07-25-11-57.opa-bundle-builder.tracing-rs.json

```

When more than 6 log files exist, the oldest one is deleted, and Vector stops watching it:

```
Found new file to watch. file=/stackable/log/bundle-builder/2024-07-25-11-58.opa-bundle-builder.tracing-rs.json
Stopped watching file. file=/stackable/log/bundle-builder/2024-07-25-11-52.opa-bundle-builder.tracing-rs.json reached_eof="true"
```

This is even simpler than with the other logging frameworks which rename the log files on roll-over.

Comment (nightkr) [reply crates/stackable-operator/src/logging/mod.rs:44]: Oh right, I mixed up `prefix` and `suffix` in the original implementation >.<

But sure, if it works I'm happy to change the format.

**Comment by lfrancke** [comment]
I just browsed through the comments here but: No rotation by size is a pretty big blocker for me. Is that really the case?

**Comment by nightkr** [comment]
Currently, yes. :/

Looks like https://github.com/tokio-rs/tracing/pull/2497 is seeing some movement again, but it's still ultimately blocked on upstream tracing actually merging it. Or we could fork it ourselves, technically (sliiightly less of a mess than h2, since this is a direct dependency rather than a transitive one, but still).

**Comment by lfrancke** [comment]
Hmmm....this is _only_ for operators though, correct?

I was thinking of https://github.com/stackabletech/issues/issues/574 but for the operators I'm way less worried. Still not great.

**Review by siegfriedweber — APPROVED** [review]
LGTM if no one vetoes it because of size-based rolling.

**Comment by lfrancke** [comment]
No veto from me

**Comment by nightkr** [comment]
> Hmmm....this is only for operators though, correct?

Operators and sidecars (like the OPA bundle builder).

**Comment by sbernauer** [comment]
I also wouldn't veto this, but have a somewhat bad feeling about this. The log rate of the bunde-builder is not that high, so it *should* be fine.

My bad feeling cat be much improved by having it running for some hours (e.g. in the e2e-security-demo), collect some basic usage numbers and do some napkin math if the emptyDir is big enough ;)

**Comment by nightkr** [comment]
Generally, bundle builder's log rate depends primarily on how often the regorule configmaps are modified. It should be very quiet for a demo, but I'm not sure how well that correlates to practical user environments.

**Comment by nightkr** [comment]
Well in practice I'd expect regorules to be fairly static.. After about a minute's hammering (constant writes every ~100ms) the logs grew to ~400KB. Extrapolating from that into the six allowed hourly files, that's ~144MB.

Not great, but also hopefully not a super realistic scenario.

That said, I agree that, long term, returning to size rotation would be ideal.

**Comment by nightkr** [comment]
Created #828 for rotation followup.

**Pull request merged**
